File: lib/db/src/schema/downloadHistory.ts
import { pgTable, text, timestamp, uuid, integer, bigint, index } from "drizzle-orm/pg-core";
import { usersTable } from "./users";

export const downloadHistoryTable = pgTable(
  "download_history",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    userId: uuid("user_id").references(() => usersTable.id, { onDelete: "cascade" }),
    platform: text("platform").notNull(),
    originalUrl: text("original_url").notNull(),
    videoTitle: text("video_title"),
    thumbnailUrl: text("thumbnail_url"),
    duration: integer("duration"),
    format: text("format").notNull(),
    quality: text("quality"),
    fileSize: bigint("file_size", { mode: "number" }),
    downloadedAt: timestamp("downloaded_at", { withTimezone: true }).defaultNow().notNull(),
    ipAddress: text("ip_address"),
    status: text("status").default("completed").notNull(),
  },
  (t) => ({
    userIdx: index("idx_history_user_id").on(t.userId),
    dateIdx: index("idx_history_downloaded_at").on(t.downloadedAt),
    platformIdx: index("idx_history_platform").on(t.platform),
  }),
);

export type DownloadHistory = typeof downloadHistoryTable.$inferSelect;
export type InsertDownloadHistory = typeof downloadHistoryTable.$inferInsert;

File: backend/app/models/download_history.py
import uuid
from datetime import datetime

from sqlalchemy import BigInteger, DateTime, ForeignKey, Integer, String, func
from sqlalchemy.dialects.postgresql import INET, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class DownloadHistory(Base):
    __tablename__ = "download_history"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    platform: Mapped[str] = mapped_column(String(20), nullable=False)
    original_url: Mapped[str] = mapped_column(String(2000), nullable=False)
    video_title: Mapped[str | None] = mapped_column(String(500), nullable=True)
    thumbnail_url: Mapped[str | None] = mapped_column(String(2000), nullable=True)
    duration: Mapped[int | None] = mapped_column(Integer, nullable=True)
    format: Mapped[str] = mapped_column(String(10), nullable=False)
    quality: Mapped[str | None] = mapped_column(String(20), nullable=True)
    file_size: Mapped[int | None] = mapped_column(BigInteger, nullable=True)
    downloaded_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        index=True,
    )
    ip_address: Mapped[str | None] = mapped_column(INET, nullable=True)
    status: Mapped[str] = mapped_column(String(20), default="completed", nullable=False)

    user = relationship("User", back_populates="history")
